### ISSUE TITLE
C++ dcgan example failing on generator

### ISSUE THREAD

**Issue by meganset** [pytorch/pytorch#30931]
## 🐛 Bug

Running the c++ example fails when running the forward calculation on the generator:

`what():  Calculated padded input size per channel: (1 x 1). Kernel size: (4 x 4). Kernel size can't be greater than actual input size (check_shape_forward at /pytorch/aten/src/ATen/native/Convolution.cpp:436)`

Is this due to recent changes to use NNPACK for strided convolutions..?

(There's also a warning about batchnorm being deprecated,
but that's a simpler fix.)

Thanks

cc @yf225

**Comment by ID56** [comment]
I had a similar problem sometime back. Try rechecking and increasing the padding in the deconvolutions, particularly the ones near the end of your model.

**Comment by meganset** [comment]
right, the sizes can be off as you set up a model.

but in this case i'm talking about the flagship c++ example in the tutorial,
[https://pytorch.org/tutorials/advanced/cpp_frontend.html]
(the _with_bias_ option has to be changed to just _bias_)

```
torch::nn::Sequential generator(
    torch::nn::Conv2d(torch::nn::Conv2dOptions(100, 256, 4).bias(false).transposed(true)),
    torch::nn::BatchNorm(256),
    torch::nn::Functional(torch::relu),
    torch::nn::Conv2d(torch::nn::Conv2dOptions(256, 128, 3).stride(2).padding(1).bias(false).transposed(true)),
    torch::nn::BatchNorm(128),
    torch::nn::Functional(torch::relu),
    torch::nn::Conv2d(torch::nn::Conv2dOptions(128, 64, 4).stride(2).padding(1).bias(false).transposed(true)),
    torch::nn::BatchNorm(64),
    torch::nn::Functional(torch::relu),
    torch::nn::Conv2d(torch::nn::Conv2dOptions(64, 1, 4).stride(2).padding(1).bias(false).transposed(true)),
    torch::nn::Functional(torch::tanh));

  generator(torch::randn({64,100,1,1}));
```
The above now fails with:
```
  what():  Calculated padded input size per channel: (1 x 1). Kernel size: (4 x 4). Kernel size can't be greater than actual input size (check_shape_forward at /pytorch/aten/src/ATen/native/Convolution.cpp:436)
```

What has changed?
Was it always incorrectly specified and is now being caught, or a recently introduced check overreacting?

thanks

**Comment by zhangguanheng66** [comment]
Could you provide a code snippet for reproduction?

**Comment by zhangguanheng66** [comment]
@yf225 Any ideas?

**Comment by meganset** [comment]
I think maybe the _nn::Conv2d_ module with _transposed(true)_ isn't being interpreted as a transposed convolution.

I tried to use the above specification, but with _ConvTranspose2d_ but this gets into some other new changes that require a 2nd argument be supplied to the forward call, and this gets trickier with Sequential..

**Comment by yf225** [comment]
@meganset Thanks for the investigation! Your assessment is correct that after https://github.com/pytorch/pytorch/pull/28917, an `nn::Conv2d` module with `transposed(true)` is not interpreted as a transposed convolution anymore, because the Python equivalent `torch.nn.Conv2d` module doesn't take a `transposed` parameter either.

The tutorial https://pytorch.org/tutorials/advanced/cpp_frontend.html is pinned to libtorch v1.3.1. I will open a PR to get the tutorial ready for our upcoming release v1.4.0, which will contain the changes to the behavior of `torch::nn::Conv{1,2,3}d` modules.

**Comment by meganset** [comment]
Thanks.
Could work around the current setup by explicitly using the _ConvTranspose2d_ call,
but within the _Sequential_ module, I get tripped up by the 2nd arg of output size,
even though it's optional..

**Comment by yf225** [comment]
@meganset My apologies for the bad experience - it seems that the `Sequential` module doesn't support adding modules with forward method that can take more than one argument. I think this is a limitation of the `Sequential` module and we will remove the usage of it in the tutorial.

**Comment by meganset** [comment]
ok, thanks for looking at this.
we really like the _Sequential_ module -- very useful when crossing language boundaries, e.g. build and use a model from another interpreter (not python) -- it would be a shame to lose tha ability to use deconvolutions with _Sequential_, but I guess we can write our own _ConvTranspose2d_ on top of the 1.4.0 one and have its forward method just accept a tensor, especially if most (nearly all?) transpose cases don't need output size specified.

**Comment by lzxzy** [comment]
@yf225 
Hi, when I try to run the tutorial. I met the same error . So I wonder is there any solution for this problem? Thx.

**Comment by lzxzy** [comment]
When I try to use the second approach to build the Generator Module like this:
```cpp
struct GeneratorImpl : nn::Module {
        GeneratorImpl(int kNoiseSize)
            : conv1(nn::ConvTranspose2dOptions(kNoiseSize, 256, 4)
                        .bias(false)),
                batch_norm1(256),
                conv2(nn::ConvTranspose2dOptions(256, 128, 3)
                        .stride(2)
                        .padding(1)
                        .bias(false)),
                batch_norm2(128),
                conv3(nn::ConvTranspose2dOptions(128, 64, 4)
                        .stride(2)
                        .padding(1)
                        .bias(false)),
                batch_norm3(64),
                conv4(nn::ConvTranspose2dOptions(64, 1, 4)
                        .stride(2)
                        .padding(1)
                        .bias(false))
        {
        // register_module() is needed if we want to use the parameters() method later on
        register_module("conv1", conv1);
        register_module("conv2", conv2);
        register_module("conv3", conv3);
        register_module("conv4", conv4);
        register_module("batch_norm1", batch_norm1);
        register_module("batch_norm2", batch_norm1);
        register_module("batch_norm3", batch_norm1);
        }

        torch::Tensor forward(torch::Tensor x) {
        x = torch::relu(batch_norm1(conv1(x)));
        x = torch::relu(batch_norm2(conv2(x)));
        x = torch::relu(batch_norm3(conv3(x)));
        x = torch::tanh(conv4(x));
        return x;
        }

        nn::ConvTranspose2d conv1, conv2, conv3, conv4;
        nn::BatchNorm batch_norm1, batch_norm2, batch_norm3;
        };
    TORCH_MODULE(Generator);
    Generator generator(kNoiseSize);
    generator->to(device);
```
It worked on CPU, but the GPU mode get the** error** ` Tensor for argument #2 'weight' is on CPU, but expected it to be on GPU`. I think it is because `generator->to(device)` doesn't work. I'm the new guy for cpp. So any idea for this?

**Comment by lzxzy** [comment]
Unfortunately, when I update the libtorch version it doesn't work. I think it's high relative with torch version

**Comment by yf225** [comment]
@lzxzy Could you try whether the following code works with the latest libtorch nightly? 
```cpp
struct DCGANGeneratorImpl : nn::Module {
  DCGANGeneratorImpl(int64_t noiseSize) {
    conv1 = register_module("conv1", nn::ConvTranspose2d(nn::ConvTranspose2dOptions(noiseSize, 256, 4).bias(false)));
    batch_norm1 = register_module("batch_norm1", nn::BatchNorm(256));
    conv2 = register_module("conv2", nn::ConvTranspose2d(nn::ConvTranspose2dOptions(256, 128, 3).stride(2).padding(1).bias(false)));
    batch_norm2 = register_module("batch_norm2", nn::BatchNorm(128));
    conv3 = register_module("conv3", nn::ConvTranspose2d(nn::ConvTranspose2dOptions(128, 64, 4).stride(2).padding(1).bias(false)));
    batch_norm3 = register_module("batch_norm3", nn::BatchNorm(64));
    conv4 = register_module("conv4", nn::ConvTranspose2d(nn::ConvTranspose2dOptions(64, 1, 4).stride(2).padding(1).bias(false)));
  }

  torch::Tensor forward(torch::Tensor x) {
    x = torch::relu(batch_norm1(conv1(x)));
    x = torch::relu(batch_norm2(conv2(x)));
    x = torch::relu(batch_norm3(conv3(x)));
    x = torch::tanh(conv4(x));
    return x;
  }

  nn::ConvTranspose2d conv1 = nullptr, conv2 = nullptr, conv3 = nullptr, conv4 = nullptr;
  nn::BatchNorm batch_norm1 = nullptr, batch_norm2 = nullptr, batch_norm3 = nullptr;
};

TORCH_MODULE(DCGANGenerator);

// in main function
DCGANGenerator generator(kNoiseSize);
generator->to(device);
```
I tried it on the latest libtorch nightly and it was able to train the model with GPU.

**Comment by lzxzy** [comment]
> @lzxzy Could you try whether the following code works with the latest libtorch nightly?
> 
> ```c++
> struct DCGANGeneratorImpl : nn::Module {
>   DCGANGeneratorImpl(int64_t noiseSize) {
>     conv1 = register_module("conv1", nn::ConvTranspose2d(nn::ConvTranspose2dOptions(noiseSize, 256, 4).bias(false)));
>     batch_norm1 = register_module("batch_norm1", nn::BatchNorm(256));
>     conv2 = register_module("conv2", nn::ConvTranspose2d(nn::ConvTranspose2dOptions(256, 128, 3).stride(2).padding(1).bias(false)));
>     batch_norm2 = register_module("batch_norm2", nn::BatchNorm(128));
>     conv3 = register_module("conv3", nn::ConvTranspose2d(nn::ConvTranspose2dOptions(128, 64, 4).stride(2).padding(1).bias(false)));
>     batch_norm3 = register_module("batch_norm3", nn::BatchNorm(64));
>     conv4 = register_module("conv4", nn::ConvTranspose2d(nn::ConvTranspose2dOptions(64, 1, 4).stride(2).padding(1).bias(false)));
>   }
> 
>   torch::Tensor forward(torch::Tensor x) {
>     x = torch::relu(batch_norm1(conv1(x)));
>     x = torch::relu(batch_norm2(conv2(x)));
>     x = torch::relu(batch_norm3(conv3(x)));
>     x = torch::tanh(conv4(x));
>     return x;
>   }
> 
>   nn::ConvTranspose2d conv1 = nullptr, conv2 = nullptr, conv3 = nullptr, conv4 = nullptr;
>   nn::BatchNorm batch_norm1 = nullptr, batch_norm2 = nullptr, batch_norm3 = nullptr;
> };
> 
> TORCH_MODULE(DCGANGenerator);
> 
> // in main function
> DCGANGenerator generator(kNoiseSize);
> generator->to(device);
> ```
> 
> I tried it on the latest libtorch nightly and it was able to train the model with GPU.

Yeah, it worked! I think maybe it's because I used the error version. THX.

**Comment by yf225** [comment]
The tutorial and its corresponding example are fixed in https://github.com/pytorch/tutorials/pull/827 and https://github.com/pytorch/examples/pull/697.

**Comment by yf225** [comment]
@meganset on the current master branch, `Sequential` can now accept module that has optional argument in its forward method (such as `ConvTranspose{1,2,3}d`) (this is fixed by https://github.com/pytorch/pytorch/pull/33027). Please feel free to try it out :D
